### PR TITLE
Get and Query API Key with profile uid

### LINKED ISSUE #98939: Return additional information about API Key creator in Get/Query API Keys responses
### Description

The [`Get API Key information API`](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-get-api-key.html) and `Query API Key API` currently returns the `username` of the API Key creator.

image

While the `username` is helpful in many contexts (particularly for self-hosted installations), it isn't all that useful for users authenticating via Elastic's Cloud SSO, as these usernames are unhelpful numeric identifiers:
image

In order to display more relevant information to our administrators, we would like to have additional fields from the creator exposed in the API response. It looks like we are capturing quite a bit of these fields already:

Larry Gregory 2023-08-28 at 08 46 47@2x

----

At a minimum, we would like to have the existing `full_name` and `email` returned in the API response.

This isn't captured in the API Key document as far as I can tell, but _ideally_, we would also like to have the creator's [User Profile](https://www.elastic.co/guide/en/elasticsearch/reference/current/user-profile.html) ID included in the response when available. The profile id would allow us to provide the most consistent experience across Kibana, as we could then leverage their preferred avatar and display name.

### PR DESCRIPTION
**Pull request by albertzaharovits**
Add new optional request option to the Get and Query API Key Information endpoints,
to return the API keys owner users' profile uid.

Closes https://github.com/elastic/elasticsearch/issues/98939

### CODE REVIEW

**Review by n1v0lg — APPROVED** [review]
LGTM. Pleasure to review!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java [hunk 1]
```diff
@@ -1955,23 +1961,22 @@ public void queryApiKeys(SearchRequest searchRequest, boolean withLimitedBy, Act
                         final long total = searchResponse.getHits().getTotalHits().value;
                         if (total == 0) {
                             logger.debug("No api keys found for query [{}]", searchRequest.source().query());
-                            listener.onResponse(QueryApiKeyResponse.EMPTY);
+                            listener.onResponse(QueryApiKeysResult.EMPTY);
                             return;
                         }
-                        final List<QueryApiKeyResponse.Item> apiKeyItem = Arrays.stream(searchResponse.getHits().getHits())
-                            .map(hit -> convertSearchHitToQueryItem(hit, withLimitedBy))
+                        List<ApiKey> apiKeyInfos = Arrays.stream(searchResponse.getHits().getHits())
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java]: Pushed [6f72925](https://github.com/elastic/elasticsearch/pull/106531/commits/6f7292578ef9197dc24b63ffefd594a0f9fa96f7)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java [hunk 2]
```diff
@@ -315,6 +323,39 @@ public void setEnabled(String uid, boolean enabled, RefreshPolicy refreshPolicy,
...
+                assert subjects.size() == resultsAndErrors.results().size();
+                Map<Subject, String> profileUidLookup = resultsAndErrors.results()
+                    .stream()
+                    .filter(t -> Objects.nonNull(t.v2()))
+                    .map(t -> new Tuple<>(t.v1(), t.v2().uid()))
+                    .collect(Collectors.toUnmodifiableMap(Tuple::v1, Tuple::v2));
+                listener.onResponse(
+                    apiKeyInfos.stream()
+                        .map(
+                            apiKeyInfo -> getApiKeyCreatorSubject(apiKeyInfo) == null
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java]: Pushed [9ad982d](https://github.com/elastic/elasticsearch/pull/106531/commits/9ad982dbfb0828a62c1aebbe76e8976fefc03483)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/QueryApiKeyResponse.java [hunk 3]
```diff
@@ -28,16 +29,44 @@
...
+            throw new IllegalArgumentException("Each api key info must be associated to a (nullable) owner profile uid");
+        }
+        int size = foundApiKeysInfos.size();
+        this.foundApiKeyInfoList = new ArrayList<>(size);
+        Iterator<ApiKey> apiKeyIterator = foundApiKeysInfos.iterator();
+        Iterator<Object[]> sortValueIterator = sortValues.iterator();
+        Iterator<String> profileUidIterator = ownerProfileUids != null ? ownerProfileUids.iterator() : null;
+        while (apiKeyIterator.hasNext()) {
+            if (false == sortValueIterator.hasNext()) {
+                throw new IllegalArgumentException("Each api key info must be associated to a (nullable) sort value");
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/QueryApiKeyResponse.java]: You're right that IAE maps to 400 and that this should instead be a case of `IllegalStateException`.
Pushed [7828535](https://github.com/elastic/elasticsearch/pull/106531/commits/7828535463dc8b227a33716317f716d96ce89bba)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/GetApiKeyResponse.java [hunk 4]
```diff
@@ -73,31 +101,56 @@ public String toString() {
...
+            return "Item{apiKeyInfo=" + apiKeyInfo + ", ownerProfileUid=" + ownerProfileUid + "}";
         }
     }
 
-    @SuppressWarnings("unchecked")
-    static final ConstructingObjectParser<GetApiKeyResponse, Void> PARSER = new ConstructingObjectParser<>(
-        "get_api_key_response",
-        args -> (args[0] == null) ? GetApiKeyResponse.EMPTY : new GetApiKeyResponse((List<ApiKey>) args[0])
-    );
+    static final ConstructingObjectParser<Item, Void> KEY_INFO_PARSER;
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/GetApiKeyResponse.java]: Pushed [5aee2e5](https://github.com/elastic/elasticsearch/pull/106531/commits/5aee2e5e28f5f1873443e6bfd6ad52fbf694ce6a)

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ApiKeyOwnerProfileIntegTests.java [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+import static org.elasticsearch.action.support.WriteRequest.RefreshPolicy.WAIT_UNTIL;
+import static org.elasticsearch.xpack.core.security.authc.support.UsernamePasswordToken.basicAuthHeaderValue;
+import static org.hamcrest.Matchers.hasItem;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.iterableWithSize;
+import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.nullValue;
+
+@ESIntegTestCase.ClusterScope(numDataNodes = 0, numClientNodes = 0, scope = ESIntegTestCase.Scope.TEST)
+public class ApiKeyOwnerProfileIntegTests extends SecurityIntegTestCase {
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ApiKeyOwnerProfileIntegTests.java:40]: Pushed [da4535c](https://github.com/elastic/elasticsearch/pull/106531/commits/da4535cdb4a5165da39a99e9b88ca20eb37ad517)

**Comment by albertzaharovits** [comment]
@elasticsearchmachine run elasticsearch-ci/8.14.0 / bwc-snapshots

**Comment by albertzaharovits** [comment]
@elasticsearchmachine run elasticsearch-ci/part-2-fips

**Pull request merged**